### PR TITLE
history_idx: do not visit a parent more than once

### PR DESCRIPTION
**Pull request by jfontan**
There are cases where it can enter a cycle

### CODE REVIEW

File: internal/function/history_idx.go [hunk 1]
```diff
@@ -153,7 +160,17 @@ func (f *HistoryIdx) repoHistoryIdx(repo *git.Repository, start, target plumbing
 		}
 
 		if c.NumParents() > 0 {
-			stack = append(stack, &stackFrame{frame.idx + 1, 0, c.ParentHashes})
+			newParents := make([]plumbing.Hash, 0, c.NumParents())
+			for _, h = range c.ParentHashes {
+				_, ok = visitedHashes[h]
+				if !ok {
```
Comment (erizocosmico) [inline internal/function/history_idx.go]: you can inline that as `if _, ok := visitedHashes[h]; !ok`

File: internal/function/history_idx.go [hunk 2]
```diff
@@ -133,7 +134,13 @@ func (f *HistoryIdx) repoHistoryIdx(repo *git.Repository, start, target plumbing
 
 		frame := stack[len(stack)-1]
 
-		c, err := repo.CommitObject(frame.hashes[frame.pos])
+		h := frame.hashes[frame.pos]
+		_, ok := visitedHashes[h]
+		if !ok {
```
Comment (erizocosmico) [inline internal/function/history_idx.go]: you can inline that as `if _, ok := visitedHashes[h]; !ok`

**Comment by ajnavarro** [comment]
The go-mysql-server broke API comp. We need another PR implementing String method and rebase here.

**Comment by erizocosmico** [comment]
Travis was failing because you were caught in the middle of making it compatible with the new go-mysql-server. Restarted the build.

**Pull request merged**